Initial implementation of GitLab statistics

nE0sIghT opened the pull request:
Issue: https://github.com/pypi/warehouse/issues/12789

di commented:
Thanks for the PR! It looks great, but I'm going to hold off on fully reviewing/merging this until #8635 is resolved for GitLab URLs, so we don't run into the same issue as we have for GitHub.

nE0sIghT commented:
#16205 looks good! I will follow it and work on Gitlab part if needed

nE0sIghT commented:
Hello!

Do I need to rebase this PR or either merge `main` branch here?

di commented:
@nE0sIghT Hang tight, I suspect the next part of https://github.com/pypi/warehouse/issues/8635 that we will resolve will be verifying the GitHub statistics, after which we can get this up to date. Thanks for your patience!

nE0sIghT commented:
Could we proceed with this PR? I can rebase or merge main branch here

di commented:
Sorry, I meant to comment here that this is unblocked now that https://github.com/pypi/warehouse/pull/16532 is merged. Please feel free to proceed!

nE0sIghT commented:
Thanks! Should I merge or rebase?

di commented:
Either is fine, we do squash merging so it doesn't make much difference.

nE0sIghT commented:
I rebased PR on top of the main branch, added unit tests and tested locally

![image](https://github.com/user-attachments/assets/0cb34e6a-0354-44a2-a9a2-bff100b38596)

di commented on warehouse/csp.py:
```diff
@@ -97,6 +97,7 @@ def _connect_src_settings(config) -> list:
         SELF,
         "https://api.github.com/repos/",
         "https://api.github.com/search/issues",
+        "https://gitlab.com",
```
```suggestion
        "https://gitlab.com/api",
```

di commented on warehouse/static/js/warehouse/controllers/gitlab_repo_stats_controller.js:
```diff
@@ -0,0 +1,82 @@
+          outlet.updateStats(stats)
+        );
+      })
+      // swallow errors, we don't want to show them to the user
+      .catch(() => {});
```
We don't want to show them to the user, but would probably be helpful to at least log them to the console here.

di commented on warehouse/static/js/warehouse/controllers/gitlab_repo_stats_controller.js:
```diff
@@ -0,0 +1,82 @@
+          );
+        }
+
+        return 0;
```
Similarly would be nice to log any errors here to the console.

di commented on warehouse/static/js/warehouse/controllers/gitlab_repo_info_controller.js:
```diff
@@ -0,0 +1,41 @@
+    "forksUrl",
+    "openIssuesCount",
+    "openIssuesUrl",
+    "stargazersCount",
+    "stargazersUrl",
```
"stargazers" seems to be a github-ism, we should just call this "followers" I think.

nE0sIghT replied:
I used "starrers" term here as in the Gitlab's UI and API

di replied:
Ah, perfect, ty!

di commented on warehouse/templates/includes/packaging/project-data.html:
```diff
@@ -92,6 +92,53 @@ <h6>{% trans %}GitHub Statistics{% endtrans %}</h6>
+        <a class="vertical-tabs__tab vertical-tabs__tab--with-icon vertical-tabs__tab--condensed"
+           data-gitlab-repo-info-target="openMRsUrl" rel="noopener">
+          <i class="fa fa-code-pull-request" aria-hidden="true"></i>
+          <strong>{% trans %}Open MRs:{% endtrans %}</strong>
```
```suggestion
          <strong>{% trans %}Open merge requests:{% endtrans %}</strong>
```

nE0sIghT commented:
Thanks for review! I pushed new commits that should address mentioned issues.

di reviewed (approved):
Thanks!

di commented:
@nE0sIghT Can you resolve the merge conflicts here? Thanks

nE0sIghT commented:
> Can you resolve the merge conflicts here?

It's done

di commented:
Sorry, we have been sorting out some issues there recently :) Yes please, should be the last time!
